iceoryx_utils/test/moduletests/test_cxx_forward_list.cpp

```diff
@@ -0,0 +1,1931 @@
+
+    iter_sut1 = cut1.before_begin();
+    iter_sut2 = cut2.before_begin();
+    EXPECT_DEATH(dummyFunc(iter_sut1 == iter_sut2), "");
+
+    iter_sut1 = cut1.end();
+    iter_sut2 = cut2.end();
+    EXPECT_DEATH(dummyFunc(iter_sut1 == iter_sut2), "");
+
+    iter_sut1 = cut1.begin();
+    iter_sut2 = cut2.begin();
+    EXPECT_DEATH(dummyFunc(iter_sut1 != iter_sut2), "");
+
+    iter_sut1 = cut1.before_begin();
+    iter_sut2 = cut2.before_begin();
+    EXPECT_DEATH(dummyFunc(iter_sut1 != iter_sut2), "");
+
+    iter_sut1 = cut1.end();
+    iter_sut2 = cut2.end();
+    EXPECT_DEATH(dummyFunc(iter_sut1 != iter_sut2), "");
+}
+
+
+TEST_F(forward_list_test, ComparingConstIteratorAndIterator)
+{
+    forward_list<CTorTest, TESTLISTCAPACITY> cut1, cut2;
+    cut1.emplace_front(15842);
+    cut1.emplace_front(1584122);
+    cut1.emplace_front(158432);
+    cut1.emplace_front(158432);
+
+    cut2.emplace_front(1313);
+    cut2.emplace_front(13131);
+
+
+    forward_list<CTorTest, TESTLISTCAPACITY>::const_iterator iter_sut1 = cut1.cbefore_begin();
+    forward_list<CTorTest, TESTLISTCAPACITY>::const_iterator iter_sut2 = cut1.cbefore_begin();
+    forward_list<CTorTest, TESTLISTCAPACITY>::iterator iter_sut3 = cut1.begin();
+    forward_list<CTorTest, TESTLISTCAPACITY>::iterator iter_sut4 = cut1.end();
+
+    ASSERT_THAT(iter_sut1 == iter_sut3, Eq(false));
+    ASSERT_THAT(iter_sut3 == iter_sut1, Eq(false));
+
+    ASSERT_THAT(iter_sut1 == iter_sut2, Eq(true));
+    ASSERT_THAT(iter_sut4 == iter_sut3, Eq(false));
+}
+
+
+TEST_F(forward_list_test, IteratorTraitsGetValueType)
+{
+    forward_list<int, 10> cut;
+
+    cut.emplace_front(5);
+    auto iter{cut.begin()};
+
+    // using a function call here is closer to the actual use case (-> intentionally did not inline all code here)
+    auto ret = iteratorTraitReturnDoubleValue(iter);
+
+    EXPECT_THAT(ret, Eq(10));
+}
+
+TEST_F(forward_list_test, IteratorTraitsCheckIteratorCategoryOnConstIterator)
+{
+    int cnt = 0;
+    auto iter = sut.cbefore_begin();
+    if (typeid(std::iterator_traits<decltype(iter)>::iterator_category) == typeid(std::random_access_iterator_tag))
+    {
+        ASSERT_TRUE(1);
+    }
+    else if (typeid(std::iterator_traits<decltype(iter)>::iterator_category) == typeid(std::forward_iterator_tag))
+    {
+        cnt = 1;
+    }
+
+    EXPECT_THAT(cnt, Eq(1));
+}
+
+TEST_F(forward_list_test, EmptyAfterClear)
+{
+    sut.emplace_front(5);
+    sut.clear();
+    EXPECT_THAT(sut.empty(), Eq(true));
+}
+
+TEST_F(forward_list_test, SizeZeroAfterClear)
+{
+    sut.emplace_front(5);
+    sut.clear();
+    EXPECT_THAT(sut.size(), Eq(0));
+}
+
+TEST_F(forward_list_test, CopyConstructor)
+{
+    forward_list<CTorTest, TESTLISTCAPACITY> cut1;
+    cut1.emplace_front(101);
+    cut1.emplace_front(102);
+    EXPECT_THAT(customCTor, Eq(2));
+
+    forward_list<CTorTest, TESTLISTCAPACITY> cut2(cut1);
+
+    EXPECT_THAT(customCTor, Eq(2));
+    EXPECT_THAT(copyCTor, Eq(2));
```

Comment:
This way we have certain expectations how many element copies are made when copying the list. This is fine, but of course restricts the way it can be implemented. Here it makes sense since it checks that the copy ctor does not copy unnecessarily.